spec/services/work_packages/set_attributes_service/derive_progress_values_status_based_spec.rb

```diff
@@ -0,0 +1,181 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+# Scenarios specified in https://community.openproject.org/wp/40749
+RSpec.describe WorkPackages::SetAttributesService::DeriveProgressValuesStatusBased,
+               type: :model,
+               with_settings: { work_package_done_ratio: "status" } do
+  shared_let(:status_0_pct_complete) { create(:status, default_done_ratio: 0, name: "0% complete") }
+  shared_let(:status_50_pct_complete) { create(:status, default_done_ratio: 50, name: "50% complete") }
+  shared_let(:status_70_pct_complete) { create(:status, default_done_ratio: 70, name: "70% complete") }
+
+  let(:user) { build_stubbed(:user) }
+  let(:project) { build_stubbed(:project) }
+  let(:work_package) { build_stubbed(:work_package, project:, status: status_0_pct_complete) }
+  let(:instance) { described_class.new(work_package) }
+
+  shared_examples_for "update progress values" do |description:|
+    subject do
+      allow(work_package)
+        .to receive(:save)
+
+      instance.call
+    end
+
+    it description do
+      work_package.attributes = set_attributes
+      all_expected_attributes = {}
+      all_expected_attributes.merge!(expected_derived_attributes) if defined?(expected_derived_attributes)
+      if defined?(expected_kept_attributes)
+        kept = work_package.attributes.slice(*expected_kept_attributes)
+        if kept.size != expected_kept_attributes.size
+          raise ArgumentError, "expected_kept_attributes contains attributes that are not present in the work_package: " \
+                               "#{expected_kept_attributes - kept.keys} not present in #{work_package.attributes}"
+        end
+        all_expected_attributes.merge!(kept)
+      end
+      next if all_expected_attributes.blank?
+
+      subject
+
+      aggregate_failures do
+        expect(work_package).to have_attributes(all_expected_attributes)
+        # work package is not saved and no errors are created
+        expect(work_package).not_to have_received(:save)
+        expect(work_package.errors).to be_empty
+      end
+    end
+  end
+
+  context "given a work package with work, remaining work, and status with % complete being set" do
+    before do
+      work_package.status = status_50_pct_complete
+      work_package.done_ratio = work_package.status.default_done_ratio
+      work_package.estimated_hours = 10.0
+      work_package.remaining_hours = 5.0
+      work_package.send(:clear_changes_information)
```

Comment:
Any specific reason to use `#send`?

Reply:
None :)

That's how it was done initially, (found it in https://github.com/opf/openproject/commit/2f9d7f408ae#diff-0e83bbd00ca6fbffb7598aaf578bdd9874deec94e1d6404e10c2b52aa8bbed61R44 and maybe it was needed at that time) and I kept carrying it over and over.

I removed it.

Good catch 😎